### PR TITLE
Add securityContext and containerPort configuration

### LINKED ISSUE #9: Nextcloud securityContext and rootlessness
# Goal
securityContext should be able to be configured in a values.yaml file.

With the configuration available, there should be default rules to secure the container to make it run as www-data instead of root (this may require running on a different port).

### PR DESCRIPTION
**Pull request by menardorama**
# Pull Request

## Description of the change

Hi,

This PR is intended to permit to run Nextcloud on a "no privileged" Kubernetes cluster. We are enforcing security by disabling container to run as root, for Nextcloud we need to rebuild the docker images in order to pass specific configuration in order to bind a non-privileged port.

This PR introduce the possibility to define securityContext (runAsNonRoot....) and change the containerPort

## Benefits

Run a Kubernetes cluster where PodSecurityPolicy (or something else like OPA) forbid to run as root.

## Possible drawbacks

From my point of view none.

## Applicable issues

- fixes #9 

## Additional information

None

## Checklist 

- [x] DCO has been [signed off on the commit](https://docs.github.com/en/github/authenticating-to-github/signing-commits).
- [x] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [ ] (optional) Variables are documented in the README.md

### CODE REVIEW

**Comment by tvories** [comment]
@menardorama if we can get this rebased I'll approve it.

**Comment by menardorama** [comment]
OK I give up with all the checks.... take it or deny it

**Comment by tvories** [comment]
Looks like you still need to sign off for the DCO and then bump the chart version to `2.6.4`.

**Comment by menardorama** [comment]
@tvories I think that everything seems to be ok now.
Can you review my PR ?

**Comment by menardorama** [comment]
I have changed chart version to current 2.6.3 as it seems that there is conflict

File: charts/nextcloud/templates/deployment.yaml [hunk 1]
```diff
@@ -371,8 +409,27 @@ spec:
...
+        {{- toYaml . | nindent 10 }}
+        {{- end }}
+        {{- end }}
       {{- else }}
       # Will mount configuration files as www-data (id: 33) for nextcloud
       securityContext:
         fsGroup: 33
+        {{- if .Values.securityContext }}
+        {{- with .Values.securityContext }}
+        {{- toYaml . | nindent 10 }}
```
Comment (saretter) [inline charts/nextcloud/templates/deployment.yaml:423]: Indentation of 10 is wrong should probably be 8.

File: charts/nextcloud/templates/deployment.yaml [hunk 2]
```diff
@@ -371,8 +409,27 @@ spec:
       # Will mount configuration files as www-data (id: 82) for nextcloud
       securityContext:
         fsGroup: 82
+        {{- if .Values.securityContext }}
+        {{- with .Values.securityContext }}
+        {{- toYaml . | nindent 10 }}
```
Comment (saretter) [inline charts/nextcloud/templates/deployment.yaml:414]: Indentation of 10 is wrong should probably be 8.

**Review by saretter — CHANGES_REQUESTED** [review]
nindent 8 is probably correct.

File: charts/nextcloud/templates/deployment.yaml (hunk 1 above)
Comment (menardorama) [reply charts/nextcloud/templates/deployment.yaml]: fixed

File: charts/nextcloud/templates/deployment.yaml (hunk 2 above)
Comment (menardorama) [reply charts/nextcloud/templates/deployment.yaml]: fixed

File: charts/nextcloud/Chart.yaml [hunk 3]
```diff
@@ -23,7 +23,7 @@ maintainers:
   email: jeff@billimek.com
 dependencies:
 - name: postgresql
-  version: 9.8.9
+  version: 10.3.7
```
Comment (tvories) [inline charts/nextcloud/Chart.yaml:26]: I think we should handle the postgresql version bump on another PR unless the version bump is required/related to the securityContext

File: charts/nextcloud/Chart.yaml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v2
 name: nextcloud
-version: 2.5.15
+version: 2.6.2
```
[the remark this answers is not in the document]
Comment (tvories) [reply charts/nextcloud/Chart.yaml]: This still needs to be bumped to 2.6.4.  The current chart version is 2.6.3 @menardorama

File: charts/nextcloud/Chart.yaml (hunk 3 above)
Comment (menardorama) [reply charts/nextcloud/Chart.yaml:26]: I understand, we forked the nextcloud helm chart on our side to handle securityContext and made some updates on postgres. 

securityContext was already handle in the version defined on 9.8.9.

I'll rollback that part

**Comment by tvories** [comment]
So close @menardorama .  Thanks for all of your work on this.  There is one final conflict that needs to be resolved and we can merge (the chart version).

**Comment by menardorama** [comment]
same player play again :-)

**Comment by tvories** [comment]
@menardorama oh no, we almost had it!  The conflict was resolved by setting the version back to 2.6.3, but it has to be 2.6.4 in order to bump the release. We just needed to resolve the conflict and accept `2.6.4` as the correct version.  One more try?

**Comment by menardorama** [comment]
@tvories ok.... done

**Comment by tvories** [comment]
@menardorama done!  Thanks for the patience and the contribution!

**Comment by BobyMCbobs** [comment]
This is wonderful, thank you @menardorama!

**Comment by menardorama** [comment]
No problem, we are happy that's finally merged. We don't have to maintain a fork on our side

**Comment by menardorama** [comment]
It may need another PR in order to add more documentation.

Apache or nginx config should be modified in order to bind non privileged port.

I can provide it, we succeed on both apache or nginx.

Feel Free to ask

**Comment by menardorama** [comment]
For example on nginx it can be defined in `values.yaml`

```
nginx:
  ## You need to set an fpm version of the image for nextcloud if you want to use nginx!
  enabled: false
  image:
    repository: nginx
    tag: alpine
    pullPolicy: IfNotPresent

  config:
    # This generates the default nginx config as per the nextcloud documentation
    default: false
    custom: |-
      server {
        listen       8080;
        server_name  localhost;

        root   /usr/share/nginx/html;
        port_in_redirect off;
        location / {
            try_files $uri $uri/ /index.html =404;
            proxy_set_header X-Real-IP  $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto https;
            proxy_set_header X-Forwarded-Port 443;
            proxy_set_header Host $host:$server_port;
        }
      }
```

For apache several files have to be changed. In my company we were just overwriting the config files by rebuilding the apache docker image (as we need mode packages it was not heavy....

**Pull request merged**